src/core/border_router/routing_manager.cpp

```diff
@@ -828,128 +828,76 @@ Error RoutingManager::SendRouterSolicitation(void)
-
-        pio = reinterpret_cast<Ip6::Nd::PrefixInfoOption *>(buffer + bufferLength);
-
-        pio->Init();
-        pio->SetOnLinkFlag();
-        pio->SetAutoAddrConfigFlag();
-        pio->SetValidLifetime(TimeMilli::MsecToSec(mOnLinkPrefixDeprecateTimer.GetFireTime() - TimerMilli::GetNow()));
+        if (mOnLinkPrefixDeprecateTimer.IsRunning())
+        {
+            validLifetime     = TimeMilli::MsecToSec(mOnLinkPrefixDeprecateTimer.GetFireTime() - TimerMilli::GetNow());
+            preferredLifetime = 0;
+        }
 
-        // Set zero preferred lifetime to immediately deprecate the advertised on-link prefix.
-        pio->SetPreferredLifetime(0);
-        pio->SetPrefix(mLocalOnLinkPrefix);
+        SuccessOrAssert(raMsg.AppendPrefixInfoOption(mLocalOnLinkPrefix, validLifetime, preferredLifetime));
 
-        bufferLength += pio->GetSize();
+        if (mIsAdvertisingLocalOnLinkPrefix)
+        {
+            mTimeAdvertisedOnLinkPrefix = TimerMilli::GetNow();
+        }
 
-        LogInfo("Send on-link prefix %s in PIO (preferred lifetime = %u seconds, valid lifetime = %u seconds)",
-                mLocalOnLinkPrefix.ToString().AsCString(), pio->GetPreferredLifetime(), pio->GetValidLifetime());
+        LogInfo("RouterAdvert: Added PIO for %s (valid=%u, preferred=%u)", mLocalOnLinkPrefix.ToString().AsCString(),
+                validLifetime, preferredLifetime);
     }
 
-    // Invalidate the advertised OMR prefixes if they are no longer in the new OMR prefix array.
+    // Invalidate previously advertised OMR prefixes if they are no
+    // longer in the new OMR prefix array.
 
-    for (const OmrPrefix &advertisedOmrPrefix : mAdvertisedOmrPrefixes)
+    for (const OmrPrefix &omrPrefix : mAdvertisedOmrPrefixes)
     {
-        if (!aNewOmrPrefixes.ContainsMatching(advertisedOmrPrefix.GetPrefix()))
+        if (!aNewOmrPrefixes.ContainsMatching(omrPrefix.GetPrefix()))
         {
-            Ip6::Nd::RouteInfoOption *rio;
-
-            OT_ASSERT(bufferLength +
-                          Ip6::Nd::RouteInfoOption::OptionSizeForPrefix(advertisedOmrPrefix.GetPrefix().GetLength()) <=
-                      sizeof(buffer));
-
-            rio = reinterpret_cast<Ip6::Nd::RouteInfoOption *>(buffer + bufferLength);
+            SuccessOrAssert(
+                raMsg.AppendRouteInfoOption(omrPrefix.GetPrefix(), /* aRouterLifetime */ 0, omrPrefix.GetPreference()));
 
-            // Set zero route lifetime to immediately invalidate the advertised OMR prefix.
-            rio->Init();
-            rio->SetRouteLifetime(0);
-            rio->SetPrefix(advertisedOmrPrefix.GetPrefix());
-
-            bufferLength += rio->GetSize();
-
-            LogInfo("Stop advertising OMR prefix %s on %s", advertisedOmrPrefix.ToString().AsCString(),
-                    mInfraIf.ToString().AsCString());
+            LogInfo("RouterAdvert: Added RIO for %s (lifetime:0)", omrPrefix.ToString().AsCString());
```

Comment:
nit:
```suggestion
            LogInfo("RouterAdvert: Added RIO for %s (lifetime=0)", omrPrefix.ToString().AsCString());
```

We are using `=` in PIO logs above,

Reply:
Changed in new push. Thanks.